### PR TITLE
CORTX_30573_TEST_39909_39910_39911

### PR DESCRIPTION
**Pull request by ankita-93**
Signed-off-by: ankita-93 <ankita.thummar@seagate.com>

# Problem Statement
- TA for TEST-39909, 39910, 39911 under [https://jts.seagate.com/browse/CORTX-30573]

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [ ] Attach test execution logs
-  [ ] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: tests/ha/test_server_pod_failure.py [hunk 1]
```diff
@@ -505,3 +506,377 @@ def test_degraded_writes_safe_server_pod_shutdown(self):
...
+        """
+        LOGGER.info("STARTED: Test to verify degraded deletes after unsafe server pod shutdown.")
+        wr_bucket = HA_CFG["s3_bucket_data"]["no_buckets_for_deg_deletes"]      # 150 Buckets
+        del_bucket = wr_bucket - 10
+        event = threading.Event()
+        wr_output = Queue()
+        del_output = Queue()
+        LOGGER.info("Step 1: Create %s buckets & perform WRITEs with variable size objects.",
+                    wr_bucket)
+        LOGGER.info("Create IAM account with name %s", self.s3acc_name)
```
Comment (ApurwaMohite) [inline tests/ha/test_server_pod_failure.py:527]: IAM user*

File: tests/ha/test_server_pod_failure.py [hunk 2]
```diff
@@ -505,3 +506,377 @@ def test_degraded_writes_safe_server_pod_shutdown(self):
...
+                                               passwd=S3_CFG["CliConfig"]["s3_account"]["password"])
+        assert_utils.assert_true(resp[0], resp[1])
+        access_key = resp[1]["access_key"]
+        secret_key = resp[1]["secret_key"]
+        self.test_prefix = 'test-39909'
+        self.s3_clean = {'s3_acc': {'accesskey': access_key, 'secretkey': secret_key,
+                                    'user_name': self.s3acc_name}}
+        s3_test_obj = S3TestLib(access_key=access_key, secret_key=secret_key,
+                                endpoint_url=S3_CFG["s3_url"])
+        LOGGER.info("Successfully created IAM account with name %s", self.s3acc_name)
```
Comment (ApurwaMohite) [inline tests/ha/test_server_pod_failure.py:539]: IAM user*

File: tests/ha/test_server_pod_failure.py [hunk 3]
```diff
@@ -505,3 +506,377 @@ def test_degraded_writes_safe_server_pod_shutdown(self):
...
+    def test_deletes_during_server_pod_down(self):
+        """
+        Following test tests continuous DELETEs during server pod down by deleting deployment
+        """
+        LOGGER.info(
+            "STARTED: Test to verify DELETEs during server pod down by deleting deployment.")
+        event = threading.Event()  # Event to be used to send when data pod restart start
+        wr_output = Queue()
+        del_output = Queue()
+        wr_bucket = HA_CFG["s3_bucket_data"]["no_buckets_for_deg_deletes"]      # 150 buckets
```
Comment (ApurwaMohite) [inline tests/ha/test_server_pod_failure.py:665]: Unnecessary comment.

File: tests/ha/test_server_pod_failure.py [hunk 4]
```diff
@@ -505,3 +506,377 @@ def test_degraded_writes_safe_server_pod_shutdown(self):
...
+        del_bucket = wr_bucket - 10
+        LOGGER.info("Create IAM account with name %s", self.s3acc_name)
```
Comment (ApurwaMohite) [inline tests/ha/test_server_pod_failure.py:667]: IAM user*
Please check at other places.

File: tests/ha/test_server_pod_failure.py [hunk 5]
```diff
@@ -505,3 +506,377 @@ def test_degraded_writes_safe_server_pod_shutdown(self):
...
+        assert_utils.assert_true(resp[0], resp[1])
+        self.s3_clean.pop('s3_acc')
+
+        LOGGER.info("Step 8: Create IAM user with multiple buckets and run IOs when cluster has "
+                    "some failures due to server pod %s going down.", pod_name)
+        users = self.mgnt_ops.create_account_users(nusers=1)
+        self.s3_clean.update(users)
+        self.test_prefix = 'test-39910-1'
+        resp = self.ha_obj.ha_s3_workload_operation(s3userinfo=list(users.values())[0],
+                                                    log_prefix=self.test_prefix, skipcleanup=True,
```
Comment (ApurwaMohite) [inline tests/ha/test_server_pod_failure.py:773]: No need to skipcleanup as we expect delete working without any issue in this test.

File: tests/ha/test_server_pod_failure.py [hunk 6]
```diff
@@ -505,3 +506,377 @@ def test_degraded_writes_safe_server_pod_shutdown(self):
...
+                                                    nsamples=2, nclients=2)
+        assert_utils.assert_true(resp[0], resp[1])
+        LOGGER.info("Step 8: Successfully created IAM user with multiple buckets and ran IOs "
+                    "when cluster has some failures due to server pod %s has gone down.", pod_name)
```
Comment (ApurwaMohite) [inline tests/ha/test_server_pod_failure.py:777]: No need to mention this "when cluster has some failures" every time.

File: tests/ha/test_server_pod_failure.py [hunk 7]
```diff
@@ -505,3 +506,377 @@ def test_degraded_writes_safe_server_pod_shutdown(self):
...
+        pod_list.remove(pod_name)
+        LOGGER.info("Step 6: Check services status on remaining pods %s", pod_list)
+        resp = self.hlth_master_list[0].get_pod_svc_status(pod_list=pod_list, fail=False)
+        LOGGER.debug("Response: %s", resp)
+        assert_utils.assert_true(resp[0], resp)
+        LOGGER.info("Step 6: Services status on remaining pods are in online state.")
+
+        event.clear()
+        thread.join()
+        del_resp = tuple()
```
Comment (ApurwaMohite) [inline tests/ha/test_server_pod_failure.py:746]: If this tuple() is not giving you any issue you can add it everywhere.

File: tests/ha/test_server_pod_failure.py [hunk 8]
```diff
@@ -505,3 +506,377 @@ def test_degraded_writes_safe_server_pod_shutdown(self):
...
+        assert_utils.assert_true(resp[0], resp)
+        LOGGER.info("Step 5: Services status on remaining pods are in online state")
+
+        event.clear()
+        thread.join()
+        responses = {}
+        while len(responses) != 2:
+            responses = output.get(timeout=HA_CFG["common_params"]["60sec_delay"])
+        LOGGER.info("Step 6: Verify status for In-flight READs & WRITEs while server pod %s is "
+                    "going down should be failed/error.", pod_name)
```
Comment (ApurwaMohite) [inline tests/ha/test_server_pod_failure.py:856]: Why failed/error?

File: tests/ha/test_server_pod_failure.py [hunk 9]
```diff
@@ -505,3 +506,377 @@ def test_degraded_writes_safe_server_pod_shutdown(self):
...
+        fail_logs = list(x[1] for x in responses["fail_res"])
+        LOGGER.debug("Fail logs list: %s", fail_logs)
+        resp = self.ha_obj.check_s3bench_log(file_paths=pass_logs)
+        assert_utils.assert_false(len(resp[1]),
+                                  f"Expected all pass, But Logs which contain failures: {resp[1]}")
+        resp = self.ha_obj.check_s3bench_log(file_paths=fail_logs, pass_logs=False)
+        assert_utils.assert_true(len(resp[1]) <= len(fail_logs),
+                                 f"Logs which contain passed IOs: {resp[1]}")
+        LOGGER.info("Step 6: Verified status for In-flight READs & WRITEs while server pod is "
+                    "going down is failed/error.", pod_name)
```
Comment (ApurwaMohite) [inline tests/ha/test_server_pod_failure.py:868]: Expected %s for formatting string.

File: tests/ha/test_server_pod_failure.py [hunk 10]
```diff
@@ -505,3 +506,377 @@ def test_degraded_writes_safe_server_pod_shutdown(self):
...
+                    "some failures due to server pod %s going down.", pod_name)
+        users = self.mgnt_ops.create_account_users(nusers=1)
+        self.s3_clean.update(users)
+        self.test_prefix = 'test-39911-1'
+        resp = self.ha_obj.ha_s3_workload_operation(s3userinfo=list(users.values())[0],
+                                                    log_prefix=self.test_prefix, skipcleanup=True,
+                                                    nsamples=2, nclients=2)
+        assert_utils.assert_true(resp[0], resp[1])
+        LOGGER.info("Step 7: Successfully created IAM user with multiple buckets and ran IOs "
+                    "when cluster has some failures due to server pod %s has gone down.", pod_name)
```
Comment (ApurwaMohite) [inline tests/ha/test_server_pod_failure.py:880]: same as above.

File: tests/ha/test_server_pod_failure.py (hunk 3 above)
Comment (ankita-93) [reply tests/ha/test_server_pod_failure.py:665]: > Unnecessary comment.

Its just adding info to test.

File: tests/ha/test_server_pod_failure.py (hunk 8 above)
Comment (ankita-93) [reply tests/ha/test_server_pod_failure.py:856]: > Why failed/error?

we are expecting failures in READ/WRITE when server pod was going down. Hence Failed or error

File: tests/ha/test_server_pod_failure.py [hunk 11]
```diff
@@ -505,3 +506,375 @@ def test_degraded_writes_safe_server_pod_shutdown(self):
...
+                    "performed WRITEs with variable size objects.", wr_bucket)
+
+        LOGGER.info("Step 2: Shutdown/Delete the server pod safely by Delete Deployment (unsafe)")
+        LOGGER.info("Get server pod name to be Shutdown/Deleted")
+        pod_list = self.node_master_list[0].get_all_pods(pod_prefix=const.SERVER_POD_NAME_PREFIX)
+        pod_name = random.sample(pod_list, 1)[0]
+        hostname = self.node_master_list[0].get_pod_hostname(pod_name=pod_name)
+        LOGGER.info("Deleting server pod %s", pod_name)
+        resp = self.node_master_list[0].delete_deployment(pod_name=pod_name)
+        LOGGER.debug("Response: %s", pod_name, resp)
```
Comment (sonalk0209) [inline tests/ha/test_server_pod_failure.py:564]: pod_name not reqd

File: tests/ha/test_server_pod_failure.py [hunk 12]
```diff
@@ -505,3 +506,375 @@ def test_degraded_writes_safe_server_pod_shutdown(self):
...
+        self.restore_pod = self.deploy = True
+        self.restore_method = const.RESTORE_DEPLOYMENT_K8S
+        LOGGER.info(
+            "Step 2: Successfully shutdown/deleted server pod %s by Delete Deployment", pod_name)
+
+        LOGGER.info("Step 3: Check cluster status.")
+        resp = self.ha_obj.check_cluster_status(self.node_master_list[0])
+        assert_utils.assert_false(resp[0], resp)
+        LOGGER.info(
+            "Step 3: Cluster has some failures due to server pod %s has gone down.", pod_name)
```
Comment (sonalk0209) [inline tests/ha/test_server_pod_failure.py:578]: is this bcoz of some pylint rule?

File: tests/ha/test_server_pod_failure.py (hunk 8 above)
Comment (sonalk0209) [reply tests/ha/test_server_pod_failure.py:855]: Yes for server pods going down, in-flight IOs are expected to fail and need to be retried from s3 client.

File: tests/ha/test_server_pod_failure.py (hunk 12 above)
Comment (ankita-93) [reply tests/ha/test_server_pod_failure.py:578]: > is this bcoz of some pylint rule?

It was not fitting in one line at 577.

**Review by sonalk0209 — APPROVED** [review]

**Comment by ankita-93** [comment]
image
Collection log for #1435

**Comment by ankita-93** [comment]
Collection log snap for this PR
image

**Review by rahul-hatwar — APPROVED** [review]
LGTM

File: tests/ha/test_server_pod_failure.py (hunk 8 above)
Comment (ApurwaMohite) [reply tests/ha/test_server_pod_failure.py:855]: I was saying why both the words are added as failed/error both depict same thing. Let it be. :)

**Review by ApurwaMohite — APPROVED** [review]

**Pull request merged**